### DISCUSSION TITLE
[CUDA] Why is the compiler invoked to generate a PTX and cubin instead of simply a `fatbin`

### DISCUSSION THREAD

**Discussion by stgeke** [libocca/occa#570] [Ideas]

**Comment by kris-rowe** [comment]
- Since fat binaries can contain multiple cubins for multiple architectures, it is likely that a single cubin is smaller than a fat binary.
- Conversely, using fat binaries could potentially reduce the need to recompile cached kernels; however, we will need to introduce smarter logic into the jitting process to leverage this.
- There are also users who are looking to deploy their compiled kernels with their application without including the OKL source. This could also benefit them.

I will table this for discussion at the next [OCCA TAF meeting](https://github.com/libocca/occa-taf/blob/main/meeting-notes/2022-04-06.md).

**Reply by stgeke** [reply]
When PTX is used the device driver will JIT compile + cache a copy of the generated binary. This caused some weird issues when running in parallel on a shared file system. Note, I am not talking about OCCA's JIT compilation and cache.

**Reply by kris-rowe** [reply]
Is there a nekRS case that will trigger the issues you were seeing? 
I will try this out on one of our systems and implement a patch using fat binaries for comparison.

**Reply by stgeke** [reply]
This happened on Summit for various cases but it's a bit random and not always reproducible. I don't think the JIT caching by the NVIDIA device driver does anything special to avoid race conditions on a shared file system. That's part of the problem. You need to ensure that the JIT cache is not shared between MPI ranks.

**Reply by tcew** [reply]
iirc occa hashes on the compiler flags as well as source code. If you go with fat binary then as long as the command line arches agree with the cache then should things just work ?

**Reply by stgeke** [reply]
Right in this case the device driver will not invoke the JIT compiler, so no caching required. Obviously there can be still caching issues coming from OCCA but that's a different topic.

**Reply by tcew** [reply]
Since you brought it up, what is the latest status on OCCA caching on Summit. Is it very reliable ? and do you just default to warm up runs to populate the caches ?

**Reply by kris-rowe** [reply]
Sorry, just noticing now that we [default to ptx generation](https://github.com/libocca/occa/blob/0acc68d7448844c8e94cd8aba4a29cadecc4bf70/src/occa/internal/modes/cuda/device.cpp#L331). Changing this to either fatbin or cubin shouldn't be an issue. 

As Tim pointed out, we [add the arch flag](https://github.com/libocca/occa/blob/0acc68d7448844c8e94cd8aba4a29cadecc4bf70/src/occa/internal/modes/cuda/device.cpp#L266) and hash against the arch version, so using the `-cubin` flag likely makes the most sense. 

Alternatively, to be more general, we could stop hashing against the arch, use `-fatbin`, and either stop adding the `-arch` flag&mdash;supposing that users will do this themselves&mdash;or add the `-arch all` flag.

**Reply by stgeke** [reply]
We have something robust now but we had to introduce several tweaks. No warm-up run is required anymore. I can share what we do in the next TAF meeting if that's of interest.
